### ISSUE TITLE
421 response code managment

### ISSUE THREAD

**Issue by hemmat** [postalhq/postal#1415]
Hi,
I've been looking at how Postal manages the different responses from mail servers and comparing it with an enterprise MTA (Elasticemail) and came across the 421 situations, some times Yahoo mail throttles the number of emails that it receives from a certain sender with a response code of 421 and responds with this message:

> Service not available, closing transmission channel. The server response was: 421 4.7.0 [TSS04] Messages from X.X.X.X temporarily deferred due to unexpected volume or user complaints - 4.16.55.1; see https://postmaster.verizonmedia.com/error-codes

In a situation like this which is a temporary error, the email should not be considered a hard fail and it should be retried a few times (with sleep times between) before it can be considered a hard fail. However, we've notices that Postal will consider such a case as a hard fail and won't retry sending those emails again.

Please take a look at the below line:
https://github.com/postalhq/postal/blob/e2409a663c5e8ece0f1201a40a7670734e2bf05c/lib/postal/http_sender.rb#L52

Could you please advise what Postal would do in such a situation?

**Comment by willpower232** [comment]
It looks like Postal loosely follows the standards where 4xx errors imply that the request to send the message would never complete (for whatever reason).

You're right though, in this situation 421 is supposed to roughly imply "try again later any maybe it will work wink wink" so if you're regularly encountering this problem, there would be a case for making it a soft fail but theres definitely a debate about semantics to be had.

**Comment by hemmat** [comment]
thanks for your answer, you are right the 4xx errors might never complete successfully but there is a chance that it might complete. In the case of Yahoo Mail, if you try to send more than roughly 40 emails to Yahoo Mail you'll probably get a 421 error but it doesn't necessarily mean that these emails will never be accepted, We've been using Elasticemail for some of our email sending operation and their system reties these emails for 9 times before giving up with periods of sleep between and almost all of emails that were soft failed by 421 the first time, get accepted after a few retries.

I think Postal treats the errors in a naive way and there is a lot of room for improvement here, the error handling should consider all cases for big email hosting services like Gmail, Yahoo Mail, Hotmail and etc, if you look at the link below for SMTP errors that Gmail might return you could see that there are many details about these errors which none of them could be handled correctly with current error handling done by Postal:

https://support.google.com/a/answer/3726730

https://support.google.com/a/answer/3221692

**Comment by adamcooke** [comment]
This is interesting. All 4xx codes are transient according to the standard and all 5xx codes are permanent. The Postal logic is actually the inverse of this which seems quite wrong and very much a bug. I'm not sure how we ended up treating these statuses like this.

**Comment by adamcooke** [comment]
I looked at this in a bit more detail. I noticed that the file you're linking to here is actually for how we handle mail delivery issues when the message is sent to an *HTTP* (rather than SMTP) endpoint. For HTTP, our logic is correct - 4xx statuses are considered hard failures whereas 5xx is temporary. 

In light of this, I am going to remove the bug flag here and rename the issue. 

We only treat certain 5xx statuses are fatal in our SMTP logic. I'd need to see more logs to determine why you're getting a hard failure for a 421 response.
